### PR TITLE
platform/buffer_from_wl_shm: Handle clients immediately destroying the buffer

### PR DESCRIPTION
**Pull request by RAOF**
We need to handle clients destroying the wl_buffer after commit. This is legal, as long as the backing store is not invalidated.

Unfortunately, clients can create a buffer from the wl_shm_pool and then immediately destroy the wl_shm_pool. This is also valid.

So in order to prevent the backing store from being invalidated we need to take a reference on the wl_shm_pool backing our shm_buffer.

However!

If we have a reference on the wl_shm_pool, attempts to resize it will fail. Real clients do this, so we can't universally take a reference to the pool.

Instead, we take a reference to the pool (and the buffer offset into it) when the buffer is being destroyed. This way only clients that do the commit/delete dance have their shm_pools frozen, and *those* clients will normally delete their shm_pools immediately anyway.

Fixes: #2643, Fixes: #2620
Reverts: #2621

### CODE REVIEW

**Comment by RAOF** [comment]
Marking this as no-merge as it *currently* allows a client that submits a bad buffer and then deletes it to crash Mir.

We need to add a SIGBUS handler to fix that. We can't use libwayland's wl_shm_buffer_{begin/end}, as *they* crash in the same circumstance.

File: src/platforms/common/server/buffer_from_wl_shm.cpp [hunk 1]
```diff
@@ -339,7 +355,16 @@ class WlShmBuffer :
...
     {
         ShmBuffer::bind();
-        notify_consumed();
+        std::lock_guard lk{upload_mutex};
+        if (!uploaded)
+        {
+            auto mapping = map_readable();
+            upload_to_texture(mapping->data(), stride_);
+            glFinish();    /* Here to ensure that all contexts see the texture (fixes multi-head rendering)
+                            * TODO: Replace with fences
```
Comment (AlanGriffiths) [inline src/platforms/common/server/buffer_from_wl_shm.cpp:364]: Note there is a `glFinish()` in `upload_to_texture()`

**Review by AlanGriffiths — APPROVED** [review]
This looks mostly right with the caveat about badly behaved clients...

> We need to add a SIGBUS handler to fix that. We can't use libwayland's wl_shm_buffer_{begin/end}, as _they_ crash in the same circumstance.

This could be an "interesting" exercise.

Signals and multithreaded programs are problematic: any SIGBUS signal need not related to this thread. And we already have a "dance" around libwayland's SIGBUS handler and ours in `mir::run_mir()`.

**Comment by RAOF** [comment]
This should now be good to go. Miriway runs locally without (this kind of) error, wlcs passes the bad buffer tests.

**Comment by RAOF** [comment]
Argh, bollocks. This is still broken - the failure path might call `wl_resource_send_error` from a non-Wayland-mainloop thread.

**Comment by RAOF** [comment]
The WLCS test failure requires https://github.com/MirServer/wlcs/pull/253 to land

**Comment by Saviq** [comment]
Kicked to refresh from main

**Comment by Saviq** [comment]
Looks like the dispmanx platform didn't get the right treatment…

**Comment by Saviq** [comment]
Apart from the wlcs failures looks like we're pushing qemu again…

```plain
 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
/<<PKGBUILDDIR>>/tools/detect_fd_leaks.bash: line 84: 63788 Segmentation fault      (core dumped) $@ 2>&1
     63789 Done                    | detect_fd_leaks
```

**Comment by AlanGriffiths** [comment]
Weirdly, building on 22.10 I see #2527 (even with precompiled headers switched off)

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Some comments from a first pass...

This is big. On the surface, there seem to be several pieces that could be pulled out for easier review and land separately:

1. SIGBUS changes
2. new shm and shm_backing implementation & tests
3. Stuff moving to better places (e.g. `wl_format_to_mir_format()` -> `DRMFormat::as_mir_format()`)

That still leaves a lot of rework to review, but breaks off some reviewable and landable chunks.

Another thing that makes reviewing hard, is that the commits related to the above are intermingled with each other and with experiments(?) that got dropped. (E.g. `buffer_from_wl_shm.*` gets changed over the first N comments and eventually deleted).

That makes it hard to select a bunch of commits to focus on one area of change without losing track of what has and hasn't been reviewed.

**Comment by RAOF** [comment]
The wlcs failure requires https://github.com/MirServer/wlcs/pull/255 to land, so that https://github.com/MirServer/wlcs/pull/253 actually results in a new wlcs package in the PPA.

**Comment by Saviq** [comment]
Looks like one flaky test still:

```plain
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from BadBufferTest
[ RUN      ] BadBufferTest.test_truncated_shm_file
[2022-12-07 09:47:31.381529] <information> mirserver: Starting
[2022-12-07 09:47:31.381636] < - debug - > mirserver: No session management supported
[2022-12-07 09:47:31.381680] <information> VT switch key handler: No VT switching support available: MinimalConsoleServices does not support VT switching
[2022-12-07 09:47:31.381783] <information> mircommon: Loading modules from: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/
[2022-12-07 09:47:31.381850] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-gbm-kms.so.20
[2022-12-07 09:47:31.381873] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-wayland.so.20
[2022-12-07 09:47:31.381887] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-eglstream-kms.so.20
[2022-12-07 09:47:31.381901] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/server-x11.so.20
[2022-12-07 09:47:31.381915] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/input-evdev.so.8
[2022-12-07 09:47:31.382661] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-throw.so
[2022-12-07 09:47:31.382701] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-dummy.so
[2022-12-07 09:47:31.382719] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/input-stub.so
[2022-12-07 09:47:31.382789] <information> mirserver: Found display driver: mir:stub-graphics (version 2.10.0)
[2022-12-07 09:47:31.382808] <information> mirserver: Driver supports:
[2022-12-07 09:47:31.382820] <information> mirserver: 	System (priority 1)
[2022-12-07 09:47:31.382833] < -warning- > mirserver: Manually-specified display platform mir:stub-graphics does not claim to support this system. Trying anyway...
[2022-12-07 09:47:31.382900] <information> mirserver: Selected display driver: mir:stub-graphics (version 2.10.0)
[2022-12-07 09:47:31.383041] <information> mircommon: Loading modules from: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/
[2022-12-07 09:47:31.383103] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-gbm-kms.so.20
[2022-12-07 09:47:31.383121] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-wayland.so.20
[2022-12-07 09:47:31.383135] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-eglstream-kms.so.20
[2022-12-07 09:47:31.383148] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/server-x11.so.20
[2022-12-07 09:47:31.383161] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/input-evdev.so.8
[2022-12-07 09:47:31.383836] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-throw.so
[2022-12-07 09:47:31.383869] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/graphics-dummy.so
[2022-12-07 09:47:31.383886] <information> mircommon: Loading module: /<<PKGBUILDDIR>>/build-amd64/lib/server-modules/input-stub.so
[2022-12-07 09:47:31.383929] <information> mirserver: Found rendering driver: mir:stub-graphics (version 2.10.0)
[2022-12-07 09:47:31.383944] <information> mirserver: Driver supports:
[2022-12-07 09:47:31.383957] <information> mirserver: 	System (priority 1)
[2022-12-07 09:47:31.383970] < -warning- > mirserver: Manually-specified rendering platform mir:stub-graphics does not claim to support this system. Trying anyway...
[2022-12-07 09:47:31.384026] <information> mirserver: Selected rendering driver: mir:stub-graphics (version 2.10.0)
[2022-12-07 09:47:31.387174] <information> mirserver: Using software cursor
[2022-12-07 09:47:31.387346] < - ERROR - > : failed to create BasicScreenShooter: ../../../../src/renderers/gl/renderer.cpp(237): Throw in function static GLuint mir::renderer::gl::Renderer::ProgramFactory::compile_shader(GLenum, const GLchar*)
Dynamic exception type: boost::wrapexcept<std::system_error>
std::exception::what: Failed to create shader: GL_NO_ERROR (0)

[2022-12-07 09:47:31.387663] <information> mirserver: Selected input driver: mir:stub-input (version: 2.10.0)
[2022-12-07 09:47:31.387719] <information> mirserver: Mir version 2.10.0
[2022-12-07 09:47:31.388661] <information> mirserver: Initial display configuration:
[2022-12-07 09:47:31.388706] <information> mirserver: * Output 1: VGA connected, used
[2022-12-07 09:47:31.388721] <information> mirserver: . |_ Physical size 0.0" 0x0mm
[2022-12-07 09:47:31.388733] <information> mirserver: . |_ Power is on
[2022-12-07 09:47:31.388746] <information> mirserver: . |_ Current mode 1600x1600 60.00Hz
[2022-12-07 09:47:31.388768] <information> mirserver: . |_ Preferred mode 1600x1600 60.00Hz
[2022-12-07 09:47:31.388780] <information> mirserver: . |_ Orientation normal
[2022-12-07 09:47:31.388792] <information> mirserver: . |_ Logical size 1600x1600
[2022-12-07 09:47:31.388803] <information> mirserver: . |_ Logical position +0+0
[2022-12-07 09:47:31.388815] <information> mirserver: . |_ Scaling factor: 1.00
./tests/test_bad_buffer.cpp:101: Failure
Failed
Expected protocol error not raised
[2022-12-07 09:47:31.394930] <information> mirserver: Stopping
[  FAILED  ] BadBufferTest.test_truncated_shm_file (19 ms)
```

File: cmake/MirCommon.cmake [hunk 2]
```diff
@@ -98,6 +98,10 @@ function (mir_discover_tests_internal EXECUTABLE TEST_ENV_OPTIONS DETECT_FD_LEAK
       set(test_exclusion_filter "${test_exclusion_filter}:AnonymousShmFile.*:MesaBufferAllocatorTest.software_buffers_dont_bypass:MesaBufferAllocatorTest.creates_software_rendering_buffer")
   endif()
 
+  if(MIR_BAD_BUFFER_TEST_ENVIRONMENT_BROKEN)
```
Comment (AlanGriffiths) [inline cmake/MirCommon.cmake:101]: Maybe `MIR_BAD_BUFFER_TEST_ENVIRONMENT_BROKEN` isn't the best name anymore?

**Comment by Saviq** [comment]
@RAOF I rebased on main and (_I think_) solved the wlcs woes.

**Comment by RAOF** [comment]
To make review easier I've pulled this apart into the core SHM handling in #2744 and the follow-on implementation of `mf::WlShm`. I think I've resolved the review comments so far that apply to #2744.

**Pull request closed without merging**